src/main/java/com/iota/iri/service/milestone/impl/LatestSolidMilestoneTrackerImpl.java

```diff
@@ -191,4 +265,32 @@ private void logChange(int prevSolidMilestoneIndex) {
             messageQ.publish("lmhs %s", latestMilestoneHash);
         }
     }
+
+    /**
+     * Tries to actively repair the ledger by reverting the milestones preceding the given milestone.<br />
+     * <br />
+     * It gets called when a milestone could not be applied to the ledger state because of problems like "inconsistent
+     * balances". While this should theoretically never happen (because milestones are by definition "consistent"), it
+     * can still happen because IRI crashed or got stopped in the middle of applying a milestone or if a milestone
+     * was processed in the wrong order.<br />
+     * <br />
+     * Every time we call this method the internal {@link #repairBackoffCounter} is incremented which causes the next
+     * call of this method to repair an additional milestone. This means that whenever we face an error we first try to
+     * reset only the last milestone, then the two last milestones, then the three last milestones (and so on ...) until
+     * the problem was fixed.<br />
+     * <br />
+     * To be able to tell when the problem is fixed and the {@link #repairBackoffCounter} can be reset, we store the
+     * milestone index that caused the problem the first time we call this method.<br />
+     *
+     * @param errorCausingMilestone the milestone that failed to be applied
+     * @throws MilestoneException if we failed to reset the corrupted milestone
+     */
+    private void repairCorruptedMilestone(MilestoneViewModel errorCausingMilestone) throws MilestoneException {
+        if(repairBackoffCounter++ == 0) {
+            errorCausingMilestoneIndex = errorCausingMilestone.index();
+        }
+        for (int i = errorCausingMilestone.index(); i > errorCausingMilestone.index() - repairBackoffCounter; i--) {
+            milestoneService.resetCorruptedMilestone(i);
+        }
+    }
```

Comment:
lets say 100 is corrupt because it didnt get its snapshotIndex set correctly and 101 also approves some of the transactions that were taken into account for the balances of 100 already. 101 can therefore not be applied.

We will first reset 101 and try to reapply it - if that fails we reset 101 and 100 and try to reapply both. If that fails we try to reapply 101, 100 and 99 and try to reapply the three of them.

Depending on which transaction didnt get its snapshotIndex correctly set, we might need to go a few milestones back to "find" the one that wasnt processed correctly. sometimes milestones 101 would reference a "broken" tx from milestone 97 for example.